### PR TITLE
App load tweaks, improve error pages

### PR DESCRIPTION
**Pull request by t3chguy**
Fixes https://github.com/vector-im/riot-web/issues/792
Needs merging in lock-step with https://github.com/matrix-org/matrix-react-sdk/pull/4462

Pending work:
- [x] redirect to static pages on complete failure
- [x] unsupported browser page inc button to ignore

Incompatible browser:
image

Incompatible browser and React failed to load:
image

Invalid config.json:
image

Something went HUGELY wrong but Modernizr seems to be happy with the browser
image

### CODE REVIEW

**Comment by t3chguy** [comment]
image

File: src/async-components/structures/CompatibilityView.tsx [hunk 1]
```diff
@@ -0,0 +1,408 @@
...
+
+import * as React from "react";
+import { _t } from "matrix-react-sdk/src/languageHandler";
+
+// directly import the style here as this layer does not support rethemedex at this time so no matrix-react-sdk
+// scss variables will be accessible.
+import "../../../res/css/structures/ErrorView.scss";
+
+interface IProps {
+    onAccept: () => void;
```
Comment (turt2live) [inline src/async-components/structures/CompatibilityView.tsx:25]:
```suggestion
    onAccept(): void;
```

Comment (turt2live) [reply src/async-components/structures/CompatibilityView.tsx:25]: sorry, this was meant to be in a review but I clicked 1ms too early while moving my mouse

File: src/async-components/structures/CompatibilityView.tsx [hunk 2]
```diff
@@ -0,0 +1,408 @@
...
+            </div>
+
+            <div className="mx_HomePage_col">
+                <div className="mx_HomePage_row">
+                    <div>
+                        <h2 id="step2_heading">Use Riot on mobile</h2>
+                        <p><strong>iOS</strong> (iPhone or iPad)</p>
+                        <a href="https://itunes.apple.com/app/riot-im/id1083446067?mt=8" target="_blank"
+                           className="mx_ClearDecoration">
+                            <svg width="144px" height="48px" viewBox="0 0 120 40" version="1.1"
```
Comment (turt2live) [inline src/async-components/structures/CompatibilityView.tsx:79]: Why do we have an entire SVG here when we should be importing it?

Applies throughout.

File: src/async-components/structures/CompatibilityView.tsx [hunk 3]
```diff
@@ -0,0 +1,408 @@
...
+                            </svg>
+                        </a>
+                    </div>
+                </div>
+            </div>
+
+            <div className="mx_HomePage_row mx_Center mx_Spacer">
+                <p className="mx_Spacer">
+                    <a href="https://riot.im" target="_blank" className="mx_FooterLink">
+                        Go to Riot.im
```
Comment (turt2live) [inline src/async-components/structures/CompatibilityView.tsx:399]: Needs translation.

File: src/async-components/structures/ErrorView.tsx [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+                        {messages && messages.map(msg => <p key={msg}>
+                            { msg }
+                        </p>)}
+                    </div>
+                </div>
+            </div>
+            <div className="mx_HomePage_row mx_Center mx_Spacer">
+                <p className="mx_Spacer">
+                    <a href="https://riot.im" target="_blank" className="mx_FooterLink">
+                        Go to Riot.im
```
Comment (turt2live) [inline src/async-components/structures/ErrorView.tsx:53]: Needs translation

File: src/async-components/structures/ErrorView.tsx [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+                    </a>
+                </p>
+            </div>
+        </div>
+    </div>;
+};
+
+ErrorView.propTypes = {
```
Comment (turt2live) [inline src/async-components/structures/ErrorView.tsx:61]: We don't use propTypes for typescript components.

File: src/vector/index.html [hunk 6]
```diff
@@ -49,7 +49,7 @@
         <% }
     } %>
   </head>
-  <body style="height: 100%;" data-vector-indexeddb-worker-script="<%= htmlWebpackPlugin.files.chunks['indexeddb-worker'].entry %>">
+  <body style="height: 100%; margin: 0;" data-vector-indexeddb-worker-script="<%= htmlWebpackPlugin.files.chunks['indexeddb-worker'].entry %>">
```
Comment (turt2live) [inline src/vector/index.html:52]: a bit unclear what this change is doing, but probably fine

File: src/vector/index.ts [hunk 7]
```diff
@@ -199,9 +201,21 @@ async function start() {
...
+
 start().catch(err => {
     console.error(err);
-    if (!acceptBrowser) {
-        // TODO redirect to static incompatible browser page
-    }
+    // show the static error in an iframe to not lose any context / console data
+    // with some basic styling to make the iframe full page
+    delete document.body.style.height;
+    const iframe = document.createElement("iframe");
```
Comment (turt2live) [inline src/vector/index.ts:210]: the iframe should have a boatload of sandboxing out of caution.

File: src/vector/static/incompatible-browser.html [hunk 8]
```diff
@@ -0,0 +1,492 @@
...
+                                    </g>
+                                </g>
+                            </g>
+                        </g>
+                    </svg>
+                </a>
+                <p class="mx_Spacer"><strong>Android</strong></p>
+                <a href="https://play.google.com/store/apps/details?id=im.vector.app" target="_blank"
+                   class="mx_ClearDecoration">
+                    <svg width="162px" height="48px" viewBox="0 0 162 48" version="1.1"
```
Comment (turt2live) [inline src/vector/static/incompatible-browser.html:287]: We really need to not have these SVGs here. If we can't import them then we should at least use relative references.

File: src/vector/index.html (hunk 6 above)
Comment (t3chguy) [reply src/vector/index.html:52]: its fixing the margin around the ErrorView and CompatibilityView in Chrome which has a default margin for when our react-sdk css doesn't get loaded yet.

File: src/vector/static/incompatible-browser.html (hunk 8 above)
Comment (t3chguy) [reply src/vector/static/incompatible-browser.html:287]: We have them in the mobile_guide, its by Nad's design to be an inline page to make it as not brittle as possible.

Comment (nadonomy) [reply src/vector/static/incompatible-browser.html:287]: Yeah, the original requirement for mobile_guide was that the page was self encapsulated and portable for implementation ease. If we want to factor the SVG's out that's fine by me, but will leave it to yourselves on technical review.

**Pull request merged**